refactor: raise test coverage to 90%+

Keyrxng opened the pull request:
Resolves https://github.com/ubiquity/devpool-directory-bounties/issues/23

The reason I've done this was because I was tired of things passing when being QA'd them then breaking a few hours later.

- raised test coverage to 90%
- tested all intended scenarios
- added the `Unavailable` label only to issues that are assigned-closed but not yet merged, removing when they are reopened-unassigned
- fixed the repoUrl parsing `Opt.in`/ `Opt.out`
- fixed the statistics to work with the new setup

(would have pushed it sooner but my actions stopped worked overnight and had to open a ticket)

- https://github.com/ubq-testing/devpool-directory/issues

![image](https://github.com/ubiquity/devpool-directory/assets/106303466/e66351fb-f85f-4d31-8176-46eef84ff572)

Keyrxng committed:
chore: test branch

Keyrxng committed:
chore: move handlers into helpers

Keyrxng committed:
chore: devpool issue handler tests

Keyrxng committed:
chore: fork tests and fix old

Keyrxng committed:
chore: merge updated mock into src

Keyrxng committed:
chore: opt fixes and tests
- removed old and unneeded logs
- getProjectUrls tests
- createDevPoolIssue tests

Keyrxng committed:
chore: update statistics and getRepoUrls

Keyrxng committed:
chore: round off tests

Keyrxng committed:
chore: comment out twitter stuff

Keyrxng committed:
chore: wrong name

Keyrxng committed:
chore: back to ubiquity

Keyrxng committed:
fix: different throws for different env

Keyrxng committed:
fix: resolve for both

0x4007 commented on helpers/github.ts:
```diff
@@ -239,16 +235,32 @@ export function getSocialMediaText(issue: GitHubIssue): string {
 }
 
-export async function getProjectUrls() {
+export async function getProjectUrls(opts: typeof opt = opt) {
```
"opts" generally implies options like parameters for a function which can be misleading. I would stick with a derivative of "opt"

0x4007 commented on helpers/twitter.ts:
```diff
@@ -38,7 +38,7 @@ async function postTweet(status: string) {
 async function deleteTweet(id: string) {
```
I had no idea that this was implemented. I wonder if it actually works as expected.

gentlementlegen replied:
After looking at the code, I have a hard time to believe it does. It seems to keep track of the posted tweets through a `json` file which I don't believe is reliable or even persisted.

Keyrxng replied:
is this holding back the PR?

0x4007 reviewed (commented):
It's a very large pull which makes it more dangerous to review from mobile and approve. 

Seems decent overall. 

In order to reduce the burden on reviewers, consider cherry picking changes and being more precise with pulls to linking issues. 

For example in the past I've done similar volume of changes, but I generally would branch off and open separate pulls based on eachother. 

Also at risk of being a bit bureaucratic, if you filed issues for the problems you are seeking to solve, we could have set pricing! But proactive fixes with a focus on fast delivery is always appreciated. Thanks!

Keyrxng commented:
if review hasn't yet begun I'd willing be to attempt to split it into distinct PRs attached to respective issues if it makes things easier for everyone involved

- Issue: Raise test coverage > PR: add new test file and adjust old test files (using mocks with untouched `src/*`)
- Issue: Refactor `src/*`, ...list of fixes > PR: merge mocks into src
- Issue: Refactor statistics > PR: (I'd redo-statistics altogether so that it doesn't rely on `Unavailable` and that can stop being applied inside `getDevpoolIssues()` and removed altogether)

I think makes sense but if someone has a better suggestion I'll do that

<hr/>

The test file I've added is monstrous, I was going test by test just using `test.only/describe.only` but it should probably be split into one file per describe block or something right?

Keyrxng commented:
Also check this run out with 25min runtime lmao

- https://github.com/ubq-testing/devpool-directory/actions/runs/8460262630

They would just sit in `Status: Queued` indefinitely until I manually cancelled the run, the odd part being that it would finish execution in normal time but it would never register as having left a `Queued` state. I could verify this because I was able to access the `twitterMap.json` output file a min or after it entering a `Queued` state which only happens as the last step.

Everything was in limbo CI, Cron, push, pr. I thought it was because of changes I had made but unfortunately not so I've no idea what caused it. I opened a ticket after a couple hours went to sleep and when I woke it had been resolved/resolved itself

gentlementlegen commented on helpers/github.ts:
```diff
@@ -347,18 +361,233 @@ export async function writeTotalRewardsToGithub(statistics: Statistics) {
   }
 }
+
+export async function createDevPoolIssue(projectIssue: GitHubIssue, projectUrl: string, body: string, twitterMap: { [key: string]: string }) {
```
Would be nice to change the `{ [key: string]: string }` to something like `Record<string, string>`

gentlementlegen commented on helpers/github.ts:
```diff
@@ -347,18 +361,233 @@ export async function writeTotalRewardsToGithub(statistics: Statistics) {
+  devpoolIssue: GitHubIssue,
+  isFork: boolean
+) {
+  //
```
```suggestion
```

gentlementlegen commented on index.ts:
```diff
@@ -40,7 +35,7 @@ async function main() {
   const devpoolIssues: GitHubIssue[] = await getAllIssues(DEVPOOL_OWNER_NAME, DEVPOOL_REPO_NAME);
 
-  // Calculate total rewards from open issues
+  // // Calculate total rewards from open issues
```
```suggestion
  // Calculate total rewards from open issues
```

gentlementlegen commented on tests/twitter.test.ts:
```diff
@@ -8,41 +8,107 @@ afterEach(() => server.resetHandlers());
-      await import("../helpers/twitter");
-    }).rejects.toThrow();
+      const { default: twitterHelper } = await import("../helpers/twitter");
+      console.log(twitterHelper);
```
```suggestion
```

Keyrxng replied:
I did this for two reasons the first was occasionally it was failing and it was because it was being imported but the not doing anything with it was likely the cause of it failing and the second was to silence the unused var error.

I tried it again without doing anything with the import and it passed each time but I've replaced it with `() => twitterHelper`. 1. to stop the unused error and 2, just in case doing nothing with the import causes nothing to happen in the test

gentlementlegen reviewed (commented):
Tested locally, all worked fine, thanks for the thorough testing. A few fixes and I am good with this.

0x4007 commented:
@gentlementlegen when I make small "code style" adjustments like you did, I directly commit the suggestion (there is a button on the GitHub UI) I think that you should just commit them and we can immediately merge etc. Looks like I can't commit your suggestions for some reason though.

gentlementlegen commented:
@pavlovcik yep I tried to do it that way? I preferred that @Keyrxng double checked and was committing himself the small changes to be sure to not break anything. Will try to push them

Keyrxng committed:
chore: syntax fixes

Keyrxng commented:
> Tested locally, all worked fine, thanks for the thorough testing. A few fixes and I am good with this.

glad to hear it

Keyrxng commented:
Would be great to see this merged if it's ready to go, man's broke 😂

gentlementlegen commented on mocks/handlers.ts:
```diff
@@ -1,3 +1,4 @@
+/* eslint-disable @typescript-eslint/no-explicit-any */
```
Would preferably not have this disabled here, what is the reason?

Keyrxng replied:
resolved by typing the label

gentlementlegen commented on mocks/mock-issue-handlers.ts:
```diff
@@ -0,0 +1,187 @@
+        });
+
+        console.log(`Updated state: ${devpoolIssue.html_url} (${projectIssue.html_url})`);
+        console.log(`${value.comment}: ${devpoolIssue.html_url} (${projectIssue.html_url})`);
```
These two logs seem redundant

gentlementlegen commented on helpers/github.ts:
```diff
@@ -347,18 +362,232 @@ export async function writeTotalRewardsToGithub(statistics: Statistics) {
+  let newState: "open" | "closed" | undefined = undefined;
+
+  // then process the state changes
+  for (const [, value] of Object.entries(stateChanges)) {
```
Why not `Object.values` if the key is not used?

Keyrxng replied:
for of guarantees sequential execution as far as I understand which is preferred here

gentlementlegen replied:
Not sure to understand why `entries` would keep it and not `values`.
```JavaScript
const obj = {
  foo: "bar",
  bar: "baz",
  baz: "boz",
};

for (const [, value] of Object.entries(obj)) {
  console.log(value);
}

console.log("-------");

for (const value of Object.values(obj)) {
  console.log(value);
}
```
would output
```text
bar
baz
boz
-------
bar
baz
boz
```
Objects are somehow predictable since ES2015 (see https://stackoverflow.com/questions/5525795/does-javascript-guarantee-object-property-order/38218582#38218582) but not guaranteed. So if that is a concern it should be changed to use an array or a map.

Keyrxng replied:
my mistake, I thought you meant to use

```js
Object.values().map // or .forEach()
```

gentlementlegen commented on helpers/github.ts:
```diff
@@ -347,18 +362,232 @@ export async function writeTotalRewardsToGithub(statistics: Statistics) {
+
+  if (
+    // only if the devpool issue is closed or the project issue is closed
+    (newState == "closed" || devpoolIssue.state == "closed") &&
```
Should use `===` to avoid any type coercion

Keyrxng replied:
I read somewhere that `===` isn't necessary with TS but I went and did a bit more reading just to be sure before commenting and I was wrong to have used `==`, my mistake

gentlementlegen commented on helpers/github.ts:
```diff
@@ -347,18 +362,232 @@ export async function writeTotalRewardsToGithub(statistics: Statistics) {
+  }
+}
+
+export async function handleDevPoolIssue(
```
This function truly is massive and would benefit to be broken down into smaller parts

gentlementlegen commented on helpers/github.ts:
```diff
@@ -347,18 +362,232 @@ export async function writeTotalRewardsToGithub(statistics: Statistics) {
+      const tweetId = await twitter.postTweet(socialMediaText);
+
+      twitterMap[createdIssue.data.node_id] = tweetId?.id ?? "";
+      await writeFile("./twitterMap.json", JSON.stringify(twitterMap));
```
How can this file be persisted?

Keyrxng replied:
It could be written to GH the same way that the total-rewards are I guess? I don't think twitter is actually in use tho

gentlementlegen replied:
I don't think so either But since this file is not committed it just gets rewritten every time so I really don't get how it works.

Keyrxng replied:
it's written into the action run outputs but as far as what happens with it after the fact or what is intended for it I'm not sure either.

gentlementlegen replied:
I thought you authored that code that's why I asked. I believe this would be for another task.

Keyrxng committed:
chore: strict eq and remove unused mock

Keyrxng committed:
chore: refactor handleDevPoolIssue

Keyrxng committed:
chore: remove type any

Keyrxng commented:
there is no issue directly tied to raising the test coverage, should there be? I think so given some of the advice pav has given me but I'm not sure in this case.

Plus `Resolves https://github.com/ubiquity/devpool-directory-bounties/issues/23` is currently assigned to another user who has closed their pr a few weeks ago and I believe that issue is resolved, so I think it would make sense to have an issue for test coverage but again, I'm unsure.

Keyrxng committed:
fix: use object.values

Keyrxng commented:
I've updated the referenced issues. But again, I'm unsure what to do about either creating a new issue for these tests specifically or removing the assignee from the linked issue, assigning me and closing that with this PR? Cheers

Keyrxng commented:
Great, just waiting for final approval from yourself @0x4007. Love to see it, and ty for the review
